walk/git.go

```diff
@@ -1,200 +1,143 @@
-		if err != nil {
-			return fmt.Errorf("failed to find root relative path for %v: %w", path, err)
-		}
-
-		if !idxCache.hasPath(relPath) {
-			log.Debugf("path %s not found in git index, skipping", relPath)
-			return nil
-		}
-
-		err = filepath.Walk(path, func(path string, info fs.FileInfo, _ error) error {
-			// skip directories
-			if info.IsDir() {
-				return nil
-			}
-
-			// determine a path relative to g.root before checking presence in the git index
-			relPath, err := filepath.Rel(g.root, path)
-			if err != nil {
-				return fmt.Errorf("failed to determine a relative path for %s: %w", path, err)
-			}
-
-			if !idxCache.hasPath(relPath) {
-				log.Debugf("path %v not found in git index, skipping", relPath)
-				return nil
-			}
-
-			file := File{
-				Path:    path,
-				RelPath: relPath,
-				Info:    info,
-			}
-
-			g.stats.Add(stats.Traversed, 1)
-			g.filesCh <- &file
-			return nil
+		// execute the command in the background
+		g.eg.Go(func() error {
+			return w.CloseWithError(cmd.Run())
 		})
-		if err != nil {
-			return fmt.Errorf("failed to walk %s: %w", path, err)
-		}
-	}
-
-	return nil
-}
 
-func (g *GitReader) Read(ctx context.Context, files []*File) (n int, err error) {
-	idx := 0
+		// create a new scanner for reading the output
+		g.scanner = bufio.NewScanner(r)
+	}
 
 LOOP:
-	for idx < len(files) {
+
+	for n < len(files) {
 		select {
+
+		// exit early if the context was cancelled
 		case <-ctx.Done():
-			return 0, ctx.Err()
-		case file, ok := <-g.filesCh:
-			if !ok {
+			return n, ctx.Err()
+
+		default:
+			// read the next file
+			if g.scanner.Scan() {
+				path := filepath.Join(g.root, g.path, g.scanner.Text())
+
+				g.log.Debugf("processing file: %s", path)
+
+				info, err := os.Stat(path)
+				if os.IsNotExist(err) {
+					// the underlying file might have been removed
+					g.log.Warnf(
+						"Path %s is in the worktree but appears to have been removed from the filesystem", path,
+					)
+					continue
+				} else if err != nil {
+					return n, fmt.Errorf("failed to stat %s: %w", path, err)
+				}
+
+				files[n] = &File{
+					Path:    path,
+					RelPath: filepath.Join(g.path, g.scanner.Text()),
+					Info:    info,
+				}
+				n++
+
+			} else {
+				// nothing more to read
 				err = io.EOF
 				break LOOP
 			}
-			files[idx] = file
-			idx++
 		}
 	}
 
-	return idx, err
+	return n, err
 }
 
 func (g *GitReader) Close() error {
 	return g.eg.Wait()
 }
 
-func NewGitReader(
+func newGitReader(
 	root string,
 	path string,
+	args []string,
 	statz *stats.Stats,
-	batchSize int,
 ) (*GitReader, error) {
-	repo, err := git.PlainOpen(root)
-	if err != nil {
-		return nil, fmt.Errorf("failed to open git repository: %w", err)
+	// check if the root is a git repository
+	cmd := exec.Command("git", "rev-parse", "--is-inside-work-tree")
+	cmd.Dir = root
+
+	if out, err := cmd.Output(); err != nil {
+		return nil, fmt.Errorf("failed to check if git repository is inside work tree: %w", err)
```

Comment:
This error message doesn't make sense, need to reword it.

Reply:
"failed to check if we're in a git repository"?